Add bookings

ReViZe96 opened the pull request:
1. booking process logic
2. add comments process logic
3. Adapting item and user entities for hibernate ORM
4. JPA repositories for working with items and users
5. compose.yaml and application.properties for postgres database
6. storage implementations for work with database

ReViZe96 committed:
java-shareit 0.2.0:
1. entities and enums for future booking processes
2. adapting item and user entities for hibernate ORM
3. JPA repositories for working with items and users
4. compose.yaml and application.properties for postgres database
5. storage implementations for work with database
6. refactoring of update-methods

TODO:
update-methods don't work (problems in @Transactional using)

ReViZe96 committed:
java-shareit 0.2.1:
1. booking methods main logic
2. remove NotOwnerTryEditException.java

TODO:
testing and maybe refactoring of booking logic
update-methods don't work (problems in @Transactional using)

ReViZe96 committed:
java-shareit 0.2.2:
1. table for comments in schema.sql (H2) and compose.yaml (postgres)
2. add comments main logic
3. BookingDto

TODO:
1. add field with comments to response in ItemController get-methods
2. testing and refactoring of booking and add-comments logic
3. update-methods don't work (problems in @Transactional using)

ReViZe96 committed:
java-shareit 0.2.2:
1. Booking entity: fields start and end + comparable
2. BookingRequestDto, BookingResponseDto, BookingMapper
3. BookingService: request-entity and BookingFilter validation
4. BookingStorage: filter bookings by time
5. DatabaseBookingStorage: refactoring if get-methods
6. DatabaseItemStorage and DatabaseUserStorage: refactoring of update-methods
7. ItemDto and ItemMapper: mapping of lastBooking, nextBooking and comments fields

TODO:
refactoring and testing in-memory mode

ReViZe96 committed:
java-shareit 0.2.3:
1. BookingServiceImpl and ItemServiceImpl refactoring
2. Comment add field created

ReViZe96 committed:
java-shareit 0.2.3:
1. code-style check

ReViZe96 committed:
java-shareit 0.2.3:
1. additional condition for adding comment on booking:
should comment after end of booking

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/dto/BookingRequestDto.java:
```diff
@@ -0,0 +1,14 @@
+@Data
+public class BookingRequestDto {
+
+    Long itemId;
```
модификаторы доступа

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/BookingMapper.java:
```diff
@@ -0,0 +1,45 @@
+        Booking booking = new Booking();
+        booking.setStart(bookingDto.getStart());
+        booking.setEnd(bookingDto.getEnd());
+        ItemDto requestedItem = itemService.getItemById(bookingDto.getItemId());
```
запрос в ItemService должен происходить на уровне сервисов. в маппере мы только маппим одни объекты на другие. ты можешь передать объект ItemDto в параметрах этого метода bookingRequestDtoToBooking

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/InMemoryBookingStorage.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.List;
+import java.util.Optional;
+
+@Component("inMemoryBookingStorage")
```
inMemory нам уже не нужен особо в этом приложении. настраиваем стандартную связь контроллер - сервис - репозиторий - БД

andrew-larkin commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -0,0 +1,19 @@
+    private String name;
+    private String description;
+    private Boolean available;
+    private Booking lastBooking;
```
в dto не должно быть моделей-сущностей. можно использовать либо id, либо другой dto

andrew-larkin commented on src/main/java/ru/practicum/shareit/item/CommentMapper.java:
```diff
@@ -0,0 +1,36 @@
+@RequiredArgsConstructor
+public class CommentMapper {
+
+    private final ItemRepository itemRepository;
```
запросов в репозитории не должно быть в мапперах. маппер - утилитарный класс. мы туда что-то отдаем и что-то оттуда получаем

andrew-larkin commented on src/main/java/ru/practicum/shareit/item/ItemMapper.java:
```diff
@@ -1,13 +1,56 @@
+public class ItemMapper {
 
-    ItemDto itemToItemDto(Item item);
+    private final CommentRepository commentRepository;
```
удаляем репозитории

andrew-larkin commented on src/main/resources/application.properties:
```diff
@@ -7,8 +7,8 @@ logging.level.org.springframework.transaction=INFO
-#spring.datasource.url
-#spring.datasource.username
-#spring.datasource.password
+# DB connection
```
закомментированные настройки удаляем

andrew-larkin reviewed (changes requested):
Привет! Хорошая реализация нового функционала! Местами требуется внести правки, обрати внимание, что я часто не дублировал комментарии при похожих ситуациях в разных местах кода.

ReViZe96 committed:
java-shareit 0.2.4:
1. refactoring of Mappers
2. booking response DTO creating method
3. incapsulate find last/next booking and comments methods in ItemService

ReViZe96 committed:
java-shareit 0.2.4:
code style check

ReViZe96 committed:
java-shareit 0.2.5:
1. refactoring of project: remove in memory save logic

andrew-larkin reviewed (approved):
Привет! Отлично, работа принята!
